worker/codalabworker/run_manager.py

```diff
@@ -0,0 +1,302 @@
+from contextlib import closing
+import httplib
+import os
+import traceback
+import threading
+
+from bundle_service_client import BundleServiceException
+from download_util import get_target_info, get_target_path, PathException
+from file_util import gzip_file, gzip_string, read_file_section, summarize_file, tar_gzip_directory
+
+
+# TODO Should this be run factory instead if all it does is creation?
+class RunManagerBase(object):
+    """
+    Base class for classes which manages individual runs on a worker.
+    Each worker has a single run manager which it uses to create, start, resume, and monitor runs.
+
+    Different implementation of this class will execute the runs in different ways.
+    For example, one run manager may submit to the local docker socket while another submits to a managed cloud compute
+    service.
+    """
+
+    @property
+    def cpus(self):
+        """
+        :return: The total available cpus for this RunManager.
+        """
+        raise NotImplementedError
+
+    @property
+    def memory_bytes(self):
+        """
+        :return: The total available memory, in bytes, for this RunManager.
+        """
+        raise NotImplementedError
+
+    @property
+    def gpus(self):
+        """
+        :return: The total available cpus for this RunManager.
+        """
+        raise NotImplementedError
+
+    def create_run(self, bundle, bundle_path, resources):
+        """
+        Creates a new run which when started will execute the provided run bundle
+        :param bundle: the run bundle to execute
+        :param bundle_path: path on the filesystem where the bundles data is stored
+        :param resources: the resources requested for this run bundle, e.g. cpu, memory, docker image
+        :return: a new Run which will execute the provided run bundle
+        """
+        raise NotImplementedError
+
+    def serialize(self, run):
+        """
+        Serialize the run in order to persist its state.
+        This is used so that workers can pickup where they left off if they are killed.
+        :param run: the run to serialize
+        :return: a dict of the serialized data for the run
+        """
+        raise NotImplementedError
+
+    def deserialize(self, run_data):
+        """
+        Deserialize run data into a run instance.
+        It is expected that the data is from a call to RunManager.serialize(run)
+        :param run_data: the serialized run data
+        :return: a new run instance which was represented by the data
+        """
+        raise NotImplementedError
+
+    def worker_did_start(self):
+        """Provides hook which can be overwritten to do something when the worker starts"""
+        pass
+
+    def worker_will_stop(self):
+        """Provides hook which can be overwritten to do something when the worker will be stopping"""
+        pass
+
+
+class RunBase(object):
```

Comment:
Yeah, it was only because I didn't want to rename the classes too much initially. 
Really it should be something like `run.py` => `docker_run.py` and `run_manager.py` => `run.py`.

I will do that rename now that it is unlikely I will get any more conflicts.